examples/widget_impl.rs

```diff
@@ -0,0 +1,342 @@
+            return Ok(());
+        }
+        if let Event::Key(key) = event::read()? {
+            match key.code {
+                KeyCode::Char('q') | KeyCode::Esc => self.should_quit = true,
+                _ => {}
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Implement the `Widget` trait on a mutable reference to the `App` type.
+///
+/// This allows the `App` type to be rendered as a widget. The `App` type owns several other widgets
+/// that are rendered as part of the app. The `Widget` trait is implemented on a mutable reference
+/// to the `App` type, which allows this to be rendered without consuming the `App` type, and allows
+/// the sub-widgets to be mutatable.
+impl Widget for &mut App {
+    fn render(self, area: Rect, buf: &mut Buffer) {
+        let constraints = Constraint::from_lengths([1, 1, 2, 1]);
+        let [greeting, timer, squares, position] = Layout::vertical(constraints).areas(area);
+
+        // render an ephemeral greeting widget
+        Greeting::new("Ratatui!").render(greeting, buf);
+
+        // render a reference to the timer widget
+        self.timer.render(timer, buf);
+
+        // render a boxed widget containing red and blue squares
+        #[cfg(feature = "unstable-widget-ref")]
+        self.boxed_squares.render(squares, buf);
+
+        // render a mutable reference to the green square widget
+        self.green_square.render(squares, buf);
+        // Display the dynamically updated position of the green square
+        let square_position = format!("Green square is at {}", self.green_square.last_position);
+        square_position.render(position, buf);
+    }
+}
+
+/// An ephemeral greeting widget.
+///
+/// This widget is implemented on the type itself, which means that it is consumed when it is
+/// rendered. This is useful for widgets that are cheap to create, don't need to be reused, and
+/// don't need to store any state between renders. This is the simplest way to implement a widget in
+/// Ratatui, but in most cases, it is better to implement the `Widget` trait on a reference to the
+/// type, as shown in the other examples below.
+///
+/// This was the way most widgets were implemented in Ratatui before `Widget` was implemented on
+/// references in [PR #903] (merged in Ratatui 0.26.0).
+///
+/// [PR #903]: https://github.com/ratatui-org/ratatui/pull/903
+struct Greeting {
+    name: String,
+}
+
+impl Greeting {
+    fn new(name: &str) -> Self {
+        Self {
+            name: name.to_string(),
+        }
+    }
+}
+
+impl Widget for Greeting {
+    fn render(self, area: Rect, buf: &mut Buffer) {
+        let greeting = format!("Hello, {}!", self.name);
+        greeting.render(area, buf);
+    }
+}
+
+/// A timer widget that displays the elapsed time since the timer was started.
+#[derive(Debug)]
+struct Timer {
+    start: Instant,
+}
+
+impl Default for Timer {
+    fn default() -> Self {
+        Self {
+            start: Instant::now(),
+        }
+    }
+}
```

Comment:
I don't like defaults that do stuff or have side effects.

```suggestion
impl Timer {
    fn start() -> Self {
        Self {
            start: Instant::now(),
        }
    }
}
```

Reply:
Instant::now() doesn't have a side effect - it's a pure function which returns the current time.
This enables App::default(), which would otherwise need to be impl App::new().
This also isn't likely something that would trip anyone up in any way.